File: javascript/src/config/config.test.ts
// Tests for the config loader.
import { generateKeyPairSync } from 'node:crypto';
import { mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { inspect } from 'node:util';
import { afterAll, describe, expect, it } from 'vitest';
import { appMode, loadFrom, NotifyProvider, Provider, SessionBackend, TasksBackend } from './config';

function mapLookup(m: Record<string, string>) {
  return (key: string): string | undefined => m[key];
}

/** A complete, valid cloudtasks configuration (the base for negative cases). */
function fullCloudtasksEnv(): Record<string, string> {
  return {
    TASKS_BACKEND: 'cloudtasks',
    TASKS_PROJECT: 'proj',
    TASKS_LOCATION: 'us-central1',
    TASKS_QUEUE: 'agent-q',
    DISPATCH_URL: 'https://svc.run.app/internal/dispatch',
    INTERNAL_TOKEN: 'sekret',
    GITHUB_WEBHOOK_SECRET: 'hmac',
  };
}

describe('config', () => {
  it('loads defaults', () => {
    const c = loadFrom(mapLookup({}));
    expect(c.llmProvider).toBe(Provider.Ollama);
    expect(c.ollamaModel).toBe('gemma4:12b');
    expect(c.ollamaCodeModel).toBe('gemma4:26b'); // default
    expect(c.notifyProvider).toBe(NotifyProvider.Slack);
    expect(c.maxIterations).toBe(3);
    expect(c.ciTimeoutMs).toBe(90 * 60 * 1000);
    expect(c.sessionBackend).toBe(SessionBackend.Memory);
    expect(c.sqliteDsn).toBe('automation-agent.db');
    expect(c.firestoreProject).toBe('');
    expect(c.firestoreCollection).toBe('automation_agent');
    expect(c.internalToken).toBe('');
    expect(c.agentPrLabel).toBe('automation-agent'); // default
  });

  it('reads a custom agent PR label', () => {
    const c = loadFrom(mapLookup({ AGENT_PR_LABEL: 'my-bot' }));
    expect(c.agentPrLabel).toBe('my-bot');
  });

  it('reads the session backend and its settings', () => {
    const c = loadFrom(
      mapLookup({
        SESSION_BACKEND: 'firestore',
        SQLITE_DSN: 'runs.db',
        FIRESTORE_PROJECT: 'my-proj',
        FIRESTORE_COLLECTION: 'agent_runs',
        INTERNAL_TOKEN: 'secret',
      }),
    );
    expect(c.sessionBackend).toBe(SessionBackend.Firestore);
    expect(c.sqliteDsn).toBe('runs.db');
    expect(c.firestoreProject).toBe('my-proj');
    expect(c.firestoreCollection).toBe('agent_runs');
    expect(c.internalToken).toBe('secret');
  });

  it('rejects an invalid session backend', () => {
    expect(() => loadFrom(mapLookup({ SESSION_BACKEND: 'redis' }))).toThrow();
  });

  it('parses the repos list', () => {
    const c = loadFrom(mapLookup({ REPOS: ' a/b , c/d ,, e/f ' }));
    expect(c.repos).toEqual(['a/b', 'c/d', 'e/f']);
  });

  it('honours the code-model override', () => {
    const c = loadFrom(mapLookup({ OLLAMA_MODEL: 'gemma4:12b', OLLAMA_CODE_MODEL: 'gemma4:26b' }));
    expect(c.ollamaCodeModel).toBe('gemma4:26b');
    expect(c.ollamaModel).toBe('gemma4:12b');
  });

  it('falls back to GH_TOKEN, with GITHUB_TOKEN taking precedence', () => {
    // GH_TOKEN is honoured when GITHUB_TOKEN is unset, so a local gh-style env works.
    expect(loadFrom(mapLookup({ GH_TOKEN: 'gh_abc' })).githubToken).toBe('gh_abc');
    // GITHUB_TOKEN wins over GH_TOKEN.
    expect(loadFrom(mapLookup({ GITHUB_TOKEN: 'primary', GH_TOKEN: 'fallback' })).githubToken).toBe(
      'primary',
    );
  });

  it('defaults the git transport to https', () => {
    const c = loadFrom(mapLookup({}));
    expect(c.gitTransport).toBe('https');
    expect(c.gitSshKey).toBe('');
  });

  it('reads the ssh git transport and an explicit key', () => {
    const c = loadFrom(mapLookup({ GIT_TRANSPORT: 'ssh', GIT_SSH_KEY: '/home/dev/.ssh/id_ed25519' }));
    expect(c.gitTransport).toBe('ssh');
    expect(c.gitSshKey).toBe('/home/dev/.ssh/id_ed25519');
  });

  it('rejects an invalid git transport', () => {
    expect(() => loadFrom(mapLookup({ GIT_TRANSPORT: 'rsync' }))).toThrow();
  });

  it('rejects an invalid LLM provider', () => {
    expect(() => loadFrom(mapLookup({ LLM_PROVIDER: 'openai' }))).toThrow();
  });

  it('rejects an invalid notify provider', () => {
    expect(() => loadFrom(mapLookup({ NOTIFY_PROVIDER: 'discord' }))).toThrow();
  });

  it('rejects an invalid duration', () => {
    expect(() => loadFrom(mapLookup({ CI_TIMEOUT: 'soon' }))).toThrow();
  });

  it('enforces the max-iterations floor', () => {
    expect(() => loadFrom(mapLookup({ MAX_ITERATIONS: '0' }))).toThrow();
  });

  it('rejects an unparseable max-iterations', () => {
    expect(() => loadFrom(mapLookup({ MAX_ITERATIONS: 'lots' }))).toThrow();
  });

  it('parses a compound duration', () => {
    const c = loadFrom(mapLookup({ CI_TIMEOUT: '1h30m' }));
    expect(c.ciTimeoutMs).toBe(90 * 60 * 1000);
  });

  it('rejects a non-numeric PORT', () => {
    expect(() => loadFrom(mapLookup({ PORT: 'abc' }))).toThrow();
  });

  it('rejects a PORT out of range', () => {
    expect(() => loadFrom(mapLookup({ PORT: '70000' }))).toThrow();
  });
});

// --- GitHub App mode ---------------------------------------------------------
// Mirrors python/tests/test_config_app.py and go/internal/config/config_app_test.go: the
// env-var contract, App-vs-PAT mode selection, positive-id and RSA-PEM validation, the
// flattened-`\n` unescape (including the trailing-newline regression), and the empty-REPOS
// rejection. No network — throwaway keys generated in-process.

function rsaPem(pkcs1 = false): string {
  const { privateKey } = generateKeyPairSync('rsa', {
    modulusLength: 2048,
    publicKeyEncoding: { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: pkcs1 ? 'pkcs1' : 'pkcs8', format: 'pem' },
  });
  return privateKey;
}

function ecPem(): string {
  const { privateKey } = generateKeyPairSync('ec', {
    namedCurve: 'P-256',
    publicKeyEncoding: { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
  });
  return privateKey;
}

// A valid RSA key shared across cases that don't probe key parsing (keygen is slow).
const APP_PEM = rsaPem();

function appEnv(overrides: Record<string, string>): Record<string, string> {
  // The full set of vars that select App mode; REPOS is included because App mode rejects
  // an empty allow-list.
  return {
    GITHUB_APP_ID: '42',
    GITHUB_APP_INSTALLATION_ID: '99',
    GITHUB_APP_PRIVATE_KEY: APP_PEM,
    REPOS: 'acme/api',
    ...overrides,
  };
}

describe('config: github app', () => {
  const tmpRoot = mkdtempSync(join(tmpdir(), 'config-app-'));
  afterAll(() => rmSync(tmpRoot, { recursive: true, force: true }));

  let keyFileSeq = 0;
  function writeKeyFile(contents: string): string {
    const p = join(tmpRoot, `key-${keyFileSeq++}.pem`);
    writeFileSync(p, contents);
    return p;
  }

  it('stays in PAT mode when no App vars are set', () => {
    const c = loadFrom(mapLookup({ GITHUB_TOKEN: 'pat', REPOS: 'acme/api' }));
    expect(appMode(c)).toBe(false);
    expect(c.githubAppId).toBe(0);
    expect(c.githubAppInstallationId).toBe(0);
    expect(c.githubAppPrivateKeyPem).toBe('');
  });

  it('selects App mode from the full var set', () => {
    const c = loadFrom(mapLookup(appEnv({})));
    expect(appMode(c)).toBe(true);
    expect(c.githubAppId).toBe(42);
    expect(c.githubAppInstallationId).toBe(99);
    expect(c.githubAppPrivateKeyPem).toContain('-----BEGIN');
  });

  it('accepts a PKCS#1 RSA key', () => {
    const c = loadFrom(mapLookup(appEnv({ GITHUB_APP_PRIVATE_KEY: rsaPem(true) })));
    expect(appMode(c)).toBe(true);
  });

  it('reads the key from a file', () => {
    const keyPath = writeKeyFile(APP_PEM);
    const c = loadFrom(
      mapLookup(appEnv({ GITHUB_APP_PRIVATE_KEY: '', GITHUB_APP_PRIVATE_KEY_PATH: keyPath })),
    );
    expect(appMode(c)).toBe(true);
    expect(c.githubAppPrivateKeyPem).toContain('-----BEGIN');
  });

  it('unescapes a flattened (literal \\n) key', () => {
    const flattened = APP_PEM.replaceAll('\n', '\\n');
    const c = loadFrom(mapLookup(appEnv({ GITHUB_APP_PRIVATE_KEY: flattened })));
    expect(appMode(c)).toBe(true);
    expect(c.githubAppPrivateKeyPem).not.toContain('\\n'); // escaped \n restored
  });

  it('unescapes a flattened key that also has a real trailing newline (from a file)', () => {
    // A secret store can flatten newlines to literal `\n` and still append one real trailing
    // newline; the unescape must run on the escaped sequences regardless. The file path is
    // read untrimmed, so this exercises the corrected condition directly.
    const keyPath = writeKeyFile(APP_PEM.replaceAll('\n', '\\n') + '\n');
    const c = loadFrom(
      mapLookup(appEnv({ GITHUB_APP_PRIVATE_KEY: '', GITHUB_APP_PRIVATE_KEY_PATH: keyPath })),
    );
    expect(appMode(c)).toBe(true);
    expect(c.githubAppPrivateKeyPem).not.toContain('\\n');
  });

  it.each([
    ['missing app id', { GITHUB_APP_ID: '' }],
    ['missing installation', { GITHUB_APP_INSTALLATION_ID: '' }],
    ['missing key', { GITHUB_APP_PRIVATE_KEY: '' }],
    ['both key sources', { GITHUB_APP_PRIVATE_KEY_PATH: '/some/key.pem' }],
    ['zero app id', { GITHUB_APP_ID: '0' }],
    ['negative app id', { GITHUB_APP_ID: '-1' }],
    ['non-numeric app id', { GITHUB_APP_ID: 'abc' }],
    ['zero installation', { GITHUB_APP_INSTALLATION_ID: '0' }],
    ['non-numeric installation', { GITHUB_APP_INSTALLATION_ID: 'x' }],
    ['invalid pem', { GITHUB_APP_PRIVATE_KEY: 'not a pem' }],
    ['empty repos in app mode', { REPOS: '' }],
  ])('rejects %s', (_name, overrides) => {
    expect(() => loadFrom(mapLookup(appEnv(overrides)))).toThrow();
  });

  it('rejects a non-RSA (EC) key', () => {
    expect(() => loadFrom(mapLookup(appEnv({ GITHUB_APP_PRIVATE_KEY: ecPem() })))).toThrow(/RSA/);
  });

  it('rejects an unreadable key file', () => {
    expect(() =>
      loadFrom(
        mapLookup(
          appEnv({ GITHUB_APP_PRIVATE_KEY: '', GITHUB_APP_PRIVATE_KEY_PATH: '/no/such/key.pem' }),
        ),
      ),
    ).toThrow(/read GITHUB_APP_PRIVATE_KEY_PATH/);
  });

  it('masks every credential when inspected/logged', () => {
    const c = loadFrom(
      mapLookup(
        appEnv({
          GITHUB_TOKEN: 'ghp_supersecretpat',
          GITHUB_WEBHOOK_SECRET: 'webhook-shhh',
          INTERNAL_TOKEN: 'internal-shhh',
          SLACK_WEBHOOK_URL: 'https://hooks.slack.com/services/SECRETPATH',
        }),
      ),
    );
    const rendered = inspect(c);
    for (const leak of [
      'ghp_supersecretpat',
      'webhook-shhh',
      'internal-shhh',
      'SECRETPATH',
      'PRIVATE KEY',
    ]) {
      expect(rendered).not.toContain(leak);
    }
    expect(rendered).toContain('***');
    // Redaction is display-only: the real values stay readable on the object itself.
    expect(c.githubToken).toBe('ghp_supersecretpat');
    expect(c.githubAppPrivateKeyPem).toContain('PRIVATE KEY');
  });
});

describe('config: execution transport (TASKS_BACKEND)', () => {
  it('defaults to the in-process backend with a 30m deadline', () => {
    const c = loadFrom(mapLookup({}));
    expect(c.tasksBackend).toBe(TasksBackend.InProcess);
    expect(c.tasksDispatchDeadlineMs).toBe(30 * 60 * 1000);
  });

  it('rejects an unknown backend', () => {
    expect(() => loadFrom(mapLookup({ TASKS_BACKEND: 'kafka' }))).toThrow();
  });

  it('requires the full cloudtasks configuration', () => {
    // Missing everything but the backend selector.
    expect(() => loadFrom(mapLookup({ TASKS_BACKEND: 'cloudtasks' }))).toThrow();
    // Drop one required key at a time from an otherwise-valid config.
    for (const key of ['TASKS_LOCATION', 'TASKS_QUEUE', 'DISPATCH_URL', 'INTERNAL_TOKEN', 'GITHUB_WEBHOOK_SECRET']) {
      const env = fullCloudtasksEnv();
      delete env[key];
      expect(() => loadFrom(mapLookup(env)), `missing ${key}`).toThrow();
    }
  });

  it('rejects an insecure or mis-pathed DISPATCH_URL', () => {
    // DISPATCH_URL must be an absolute https URL ending in /internal/dispatch — the task carries
    // the Bearer token to it, so a plaintext / wrong-path target is rejected.
    for (const bad of [
      'http://svc.run.app/internal/dispatch', // plaintext
      '/internal/dispatch', // relative
      'not a url', // garbage
      'https://svc.run.app/', // right scheme, wrong path
    ]) {
      const env = fullCloudtasksEnv();
      env.DISPATCH_URL = bad;
      expect(() => loadFrom(mapLookup(env)), bad).toThrow();
    }
  });

  it('rejects an out-of-range dispatch deadline', () => {
    // Cloud Tasks clamps an HTTP-target dispatch deadline to 15s..30m.
    for (const bad of ['10s', '45m']) {
      const env = fullCloudtasksEnv();
      env.TASKS_DISPATCH_DEADLINE = bad;
      expect(() => loadFrom(mapLookup(env)), bad).toThrow();
    }
  });

  it('loads a full cloudtasks configuration', () => {
    const c = loadFrom(mapLookup(fullCloudtasksEnv()));
    expect(c.tasksBackend).toBe(TasksBackend.CloudTasks);
    expect(c.tasksProject).toBe('proj');
    expect(c.tasksLocation).toBe('us-central1');
    expect(c.tasksQueue).toBe('agent-q');
    expect(c.dispatchUrl).toBe('https://svc.run.app/internal/dispatch');
  });

  it('falls back to GOOGLE_CLOUD_PROJECT for the tasks project', () => {
    const env = fullCloudtasksEnv();
    delete env.TASKS_PROJECT;
    env.GOOGLE_CLOUD_PROJECT = 'ambient';
    const c = loadFrom(mapLookup(env));
    expect(c.tasksProject).toBe('ambient');
  });
});

File: kotlin/src/test/kotlin/com/automation/agent/config/ConfigTest.kt
package com.automation.agent.config

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import org.bouncycastle.asn1.pkcs.PrivateKeyInfo
import org.bouncycastle.openssl.jcajce.JcaPEMWriter
import org.bouncycastle.openssl.jcajce.JcaPKCS8Generator
import org.bouncycastle.util.io.pem.PemObject
import java.io.StringWriter
import java.nio.file.Files
import java.security.KeyPair
import java.security.KeyPairGenerator

private fun lookupOf(m: Map<String, String>): Config.Companion.Lookup =
    Config.Companion.Lookup { m[it] }

private fun rsaPair(): KeyPair = KeyPairGenerator.getInstance("RSA").apply { initialize(2048) }.generateKeyPair()

private fun pkcs8Pem(kp: KeyPair): String {
    val sw = StringWriter()
    JcaPEMWriter(sw).use { it.writeObject(JcaPKCS8Generator(kp.private, null)) }
    return sw.toString()
}

private fun pkcs1Pem(kp: KeyPair): String {
    val pkcs1 = PrivateKeyInfo.getInstance(kp.private.encoded).parsePrivateKey().toASN1Primitive().encoded
    val sw = StringWriter()
    JcaPEMWriter(sw).use { it.writeObject(PemObject("RSA PRIVATE KEY", pkcs1)) }
    return sw.toString()
}

private fun ecPem(): String {
    val kp = KeyPairGenerator.getInstance("EC").apply { initialize(256) }.generateKeyPair()
    val sw = StringWriter()
    JcaPEMWriter(sw).use { it.writeObject(JcaPKCS8Generator(kp.private, null)) }
    return sw.toString()
}

/** A throwaway RSA key in PKCS#8 PEM, reused across the App-mode cases. */
private val APP_PEM: String = pkcs8Pem(rsaPair())

/** The full App env, with [overrides] merged in. An override value of "" reads as unset (getOr). */
private fun appEnv(overrides: Map<String, String> = emptyMap()): Map<String, String> {
    val base = mutableMapOf(
        "GITHUB_APP_ID" to "42",
        "GITHUB_APP_INSTALLATION_ID" to "99",
        "GITHUB_APP_PRIVATE_KEY" to APP_PEM,
        "REPOS" to "acme/api",
    )
    base.putAll(overrides)
    return base
}

private fun writeKeyFile(contents: String): String {
    val p = Files.createTempFile("config-app-", ".pem")
    Files.writeString(p, contents)
    return p.toString()
}

class ConfigTest : BehaviorSpec({
    Given("an environment with no variables set") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(emptyMap()))
            Then("it applies the documented defaults") {
                c.llmProvider shouldBe Provider.OLLAMA
                c.ollamaModel shouldBe "gemma4:12b"
                c.ollamaCodeModel shouldBe "gemma4:26b"
                c.notifyProvider shouldBe NotifyProvider.SLACK
                c.maxIterations shouldBe 3
                c.ciTimeout.inWholeMinutes shouldBe 90L
                c.agentPrLabel shouldBe "automation-agent"
                c.sessionBackend shouldBe SessionBackend.MEMORY
                c.sqliteDsn shouldBe "automation-agent.db"
                c.firestoreProject shouldBe ""
                c.firestoreCollection shouldBe "automation_agent"
                c.internalToken shouldBe ""
                c.gitTransport shouldBe "https"
                c.gitSshKey shouldBe ""
            }
        }
    }

    Given("an ssh git transport with an explicit key") {
        When("loading the configuration") {
            val c = Config.loadFrom(
                lookupOf(
                    mapOf(
                        "GIT_TRANSPORT" to "ssh",
                        "GIT_SSH_KEY" to "/home/dev/.ssh/id_ed25519",
                    ),
                ),
            )
            Then("the transport and key path are read") {
                c.gitTransport shouldBe "ssh"
                c.gitSshKey shouldBe "/home/dev/.ssh/id_ed25519"
            }
        }
    }

    Given("an invalid GIT_TRANSPORT") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("GIT_TRANSPORT" to "scp")))
                }
            }
        }
    }

    Given("explicit session-backend settings") {
        When("loading the sqlite backend with overrides") {
            val c = Config.loadFrom(
                lookupOf(
                    mapOf(
                        "SESSION_BACKEND" to "sqlite",
                        "SQLITE_DSN" to "/data/agent.db",
                        "INTERNAL_TOKEN" to "sekret",
                    ),
                ),
            )
            Then("the backend, DSN, and internal token are read") {
                c.sessionBackend shouldBe SessionBackend.SQLITE
                c.sqliteDsn shouldBe "/data/agent.db"
                c.internalToken shouldBe "sekret"
            }
        }

        When("loading the firestore backend with overrides") {
            val c = Config.loadFrom(
                lookupOf(
                    mapOf(
                        "SESSION_BACKEND" to "firestore",
                        "FIRESTORE_PROJECT" to "my-proj",
                        "FIRESTORE_COLLECTION" to "agent_runs",
                    ),
                ),
            )
            Then("the backend, project, and collection are read") {
                c.sessionBackend shouldBe SessionBackend.FIRESTORE
                c.firestoreProject shouldBe "my-proj"
                c.firestoreCollection shouldBe "agent_runs"
            }
        }
    }

    Given("an invalid SESSION_BACKEND") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("SESSION_BACKEND" to "postgres")))
                }
            }
        }
    }

    Given("a REPOS value with surrounding whitespace and empty entries") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(mapOf("REPOS" to " a/b , c/d ,, e/f ")))
            Then("repositories are trimmed and empties dropped") {
                c.repos shouldBe listOf("a/b", "c/d", "e/f")
            }
        }
    }

    Given("an explicit OLLAMA_CODE_MODEL override") {
        When("loading the configuration") {
            val c = Config.loadFrom(
                lookupOf(mapOf("OLLAMA_MODEL" to "gemma4:12b", "OLLAMA_CODE_MODEL" to "gemma4:26b")),
            )
            Then("the code model is used and the base model is unchanged") {
                c.ollamaCodeModel shouldBe "gemma4:26b"
                c.ollamaModel shouldBe "gemma4:12b"
            }
        }
    }

    Given("GH_TOKEN set but GITHUB_TOKEN unset") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(mapOf("GH_TOKEN" to "gh_abc")))
            Then("GH_TOKEN is honoured so a local gh-style env works") {
                c.githubToken shouldBe "gh_abc"
            }
        }
    }

    Given("both GITHUB_TOKEN and GH_TOKEN set") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(mapOf("GITHUB_TOKEN" to "primary", "GH_TOKEN" to "fallback")))
            Then("GITHUB_TOKEN takes precedence") {
                c.githubToken shouldBe "primary"
            }
        }
    }

    Given("a compound CI_TIMEOUT duration") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(mapOf("CI_TIMEOUT" to "1h30m")))
            Then("it parses to the summed duration") {
                c.ciTimeout.inWholeMinutes shouldBe 90L
            }
        }
    }

    Given("an invalid LLM_PROVIDER") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("LLM_PROVIDER" to "openai")))
                }
            }
        }
    }

    Given("an invalid NOTIFY_PROVIDER") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("NOTIFY_PROVIDER" to "discord")))
                }
            }
        }
    }

    Given("an unparseable CI_TIMEOUT") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("CI_TIMEOUT" to "soon")))
                }
            }
        }
    }

    Given("a non-numeric MAX_ITERATIONS") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("MAX_ITERATIONS" to "lots")))
                }
            }
        }
    }

    Given("MAX_ITERATIONS below the floor") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("MAX_ITERATIONS" to "0")))
                }
            }
        }
    }

    Given("a non-numeric PORT") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("PORT" to "abc")))
                }
            }
        }
    }

    Given("a PORT out of range") {
        When("loading the configuration") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(mapOf("PORT" to "70000")))
                }
            }
        }
    }

    Given("no GitHub App vars") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(mapOf("GITHUB_TOKEN" to "pat", "REPOS" to "acme/api")))
            Then("it stays in PAT mode with zero App fields") {
                c.appMode() shouldBe false
                c.githubAppId shouldBe 0L
                c.githubAppInstallationId shouldBe 0L
                c.githubAppPrivateKeyPem shouldBe ""
            }
        }
    }

    Given("the full GitHub App var set") {
        When("loading the configuration") {
            val c = Config.loadFrom(lookupOf(appEnv()))
            Then("App mode is selected and the credentials are read") {
                c.appMode() shouldBe true
                c.githubAppId shouldBe 42L
                c.githubAppInstallationId shouldBe 99L
                c.githubAppPrivateKeyPem shouldContain "-----BEGIN"
            }
        }

        When("the key is a PKCS#1 RSA key") {
            val c = Config.loadFrom(lookupOf(appEnv(mapOf("GITHUB_APP_PRIVATE_KEY" to pkcs1Pem(rsaPair())))))
            Then("it is accepted") {
                c.appMode() shouldBe true
            }
        }

        When("the key is read from a file") {
            val path = writeKeyFile(APP_PEM)
            val c = Config.loadFrom(
                lookupOf(appEnv(mapOf("GITHUB_APP_PRIVATE_KEY" to "", "GITHUB_APP_PRIVATE_KEY_PATH" to path))),
            )
            Then("App mode is selected from the file contents") {
                c.appMode() shouldBe true
                c.githubAppPrivateKeyPem shouldContain "-----BEGIN"
            }
        }

        When("the key is flattened to literal \\n") {
            val c = Config.loadFrom(lookupOf(appEnv(mapOf("GITHUB_APP_PRIVATE_KEY" to APP_PEM.replace("\n", "\\n")))))
            Then("the escaped newlines are restored") {
                c.appMode() shouldBe true
                c.githubAppPrivateKeyPem shouldNotContain "\\n"
            }
        }

        When("the key is flattened AND has a real trailing newline (from a file)") {
            // A secret store can flatten newlines to literal `\n` and still append one real trailing
            // newline; the unescape must run on the escaped sequences regardless. The file path is
            // read untrimmed, exercising the corrected condition directly.
            val path = writeKeyFile(APP_PEM.replace("\n", "\\n") + "\n")
            val c = Config.loadFrom(
                lookupOf(appEnv(mapOf("GITHUB_APP_PRIVATE_KEY" to "", "GITHUB_APP_PRIVATE_KEY_PATH" to path))),
            )
            Then("the escaped newlines are still restored") {
                c.appMode() shouldBe true
                c.githubAppPrivateKeyPem shouldNotContain "\\n"
            }
        }
    }

    Given("a config holding secrets") {
        When("it is stringified (e.g. by a debug/startup log)") {
            val c = Config.loadFrom(
                lookupOf(
                    appEnv(
                        mapOf(
                            "GITHUB_TOKEN" to "ghp_supersecretpat",
                            "GITHUB_WEBHOOK_SECRET" to "webhook-shhh",
                            "INTERNAL_TOKEN" to "internal-shhh",
                            "SLACK_WEBHOOK_URL" to "https://hooks.slack.com/services/SECRETPATH",
                        ),
                    ),
                ),
            )
            val s = c.toString()
            Then("every credential value is masked and none leaks verbatim") {
                s shouldNotContain "ghp_supersecretpat"
                s shouldNotContain "webhook-shhh"
                s shouldNotContain "internal-shhh"
                s shouldNotContain "SECRETPATH"
                s shouldNotContain "-----BEGIN" // the App private key PEM must never appear
                s shouldContain "***"
                s shouldContain "githubAppId=42" // non-secret fields stay visible for debugging
            }
        }
    }

    Given("a misconfigured GitHub App env") {
        listOf(
            "missing app id" to mapOf("GITHUB_APP_ID" to ""),
            "missing installation" to mapOf("GITHUB_APP_INSTALLATION_ID" to ""),
            "missing key" to mapOf("GITHUB_APP_PRIVATE_KEY" to ""),
            "both key sources" to mapOf("GITHUB_APP_PRIVATE_KEY_PATH" to "/some/key.pem"),
            "zero app id" to mapOf("GITHUB_APP_ID" to "0"),
            "negative app id" to mapOf("GITHUB_APP_ID" to "-1"),
            "non-numeric app id" to mapOf("GITHUB_APP_ID" to "abc"),
            "zero installation" to mapOf("GITHUB_APP_INSTALLATION_ID" to "0"),
            "non-numeric installation" to mapOf("GITHUB_APP_INSTALLATION_ID" to "x"),
            "invalid pem" to mapOf("GITHUB_APP_PRIVATE_KEY" to "not a pem"),
            "empty repos in app mode" to mapOf("REPOS" to ""),
        ).forEach { (name, overrides) ->
            When("loading with $name") {
                Then("it fails") {
                    shouldThrow<IllegalArgumentException> { Config.loadFrom(lookupOf(appEnv(overrides))) }
                }
            }
        }

        When("the key is a non-RSA (EC) key") {
            Then("it is rejected as not RSA") {
                val err = shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(lookupOf(appEnv(mapOf("GITHUB_APP_PRIVATE_KEY" to ecPem()))))
                }
                err.message shouldContain "RSA"
            }
        }

        When("the key file is unreadable") {
            Then("it reports the read failure") {
                // A guaranteed-missing child of a fresh temp dir, so the read-failure branch is
                // exercised deterministically (a host-dependent literal path could exist on a runner).
                val missing = Files.createTempDirectory("config-app-missing-").resolve("missing.pem").toString()
                val err = shouldThrow<IllegalArgumentException> {
                    Config.loadFrom(
                        lookupOf(
                            appEnv(
                                mapOf(
                                    "GITHUB_APP_PRIVATE_KEY" to "",
                                    "GITHUB_APP_PRIVATE_KEY_PATH" to missing,
                                ),
                            ),
                        ),
                    )
                }
                err.message shouldContain "read GITHUB_APP_PRIVATE_KEY_PATH"
            }
        }
    }
})
